### ISSUE TITLE
Varying results when calling model.track vs. doing tracking "by foot"

### ISSUE THREAD

**Issue by jens-siebert** [ultralytics/ultralytics#20934]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hey there,

I'm trying to understand why an object in the same scene is assigned different tracking IDs depending on whether I call `model.track`, or whether I initialize the tracker manually and perform tracking via `model.predict` and `tracker.update`.

The scene contains a pallet on a pallet truck that is pulled through a so-called “gate”, a portal that is used to measure the pallet and scan barcodes attached to the pallet. The pallet is first pulled through the gate from left to right, then disappears briefly from the scene and is then pulled through the gate again from right to left.

If I track the pallet using `model.track`, the bounding box is assigned ID 1 both on the outward and return journey, even though the pallet was not visible for several frames (~90 frames).

https://github.com/user-attachments/assets/e6708931-be2c-4987-a43b-69ac5fef6835

However, if I initialize the tracker manually and perform tracking using `model.predict` and `tracker.update`, the pallet is assigned ID 2 on the way back.

https://github.com/user-attachments/assets/5a8b353c-3893-4719-ae4f-11ffc204d1a4

I would have expected this behavior in the first case as well. Is there perhaps a plausible explanation for this deviating behavior in the first case?

Please find the code I used for both cases below...

Thanks in advance!

### Additional

Here's the code which I used for the first case (using `model.track`):
```python
import os

import cv2

from ultralytics import YOLO

model_file = "palette_tracking_yolo11s_2025-04-07.onnx"
input_path = "images"
output_path = "results"

model = YOLO(model_file, task="detect", verbose=False)

image_files = []
for dirpath, dirnames, filenames in os.walk(input_path):
    for filename in filenames:
        image_file = os.path.join(dirpath, filename)
        if (os.path.isfile(image_file) and (image_file.lower().endswith(".jpg") or image_file.lower().endswith(".png"))):
            image_files.append(image_file)

image_files = sorted(list(image_files))

if not os.path.exists(output_path):
    os.makedirs(output_path)
else:
    for file in os.listdir(output_path):
        os.remove(os.path.join(output_path, file))

for img_index, img in enumerate(image_files):
    result = model.track(img, persist=True, verbose=False, conf=0.7, tracker="bytetrack.yaml")

    output_file = os.path.join(output_path, f"image_{str(img_index).zfill(6)}.jpg")
    cv2.imwrite(output_file, result[0].plot())
```

And here's the code I used for the second case (using `model.predict` and `tracker.update`):
```python
import argparse
import os

import cv2
import torch

from ultralytics import YOLO
from ultralytics.trackers import BYTETracker


model_file = "palette_tracking_yolo11s_2025-04-07.onnx"
input_path = "images"
output_path = "results"

# Same config values as "bytetrack.yaml"
tracker = BYTETracker(argparse.Namespace(
    track_high_thresh=0.25,
    track_low_thresh=0.1,
    new_track_thresh=0.25,
    track_buffer=30,
    match_thresh=0.8,
    fuse_score=True,
))

model = YOLO(model_file, task="detect", verbose=False)

image_files = []
for dirpath, dirnames, filenames in os.walk(input_path):
    for filename in filenames:
        image_file = os.path.join(dirpath, filename)
        if (os.path.isfile(image_file) and (image_file.lower().endswith(".jpg") or image_file.lower().endswith(".png"))):
            image_files.append(image_file)

image_files = sorted(list(image_files))

if not os.path.exists(output_path):
    os.makedirs(output_path)
else:
    for file in os.listdir(output_path):
        os.remove(os.path.join(output_path, file))

for img_index, img in enumerate(image_files):
    result = model.predict(img, verbose=False, conf=0.7)
    boxes = result[0].boxes.cpu().numpy()
    tracks = tracker.update(boxes, img)

    if len(tracks > 0):
        update_args = dict()
        update_args["boxes"] = torch.as_tensor(tracks[:, :-1])
        result[0].update(**update_args)

    output_file = os.path.join(output_path, f"image_{str(img_index).zfill(6)}.jpg")
    cv2.imwrite(output_file, result[0].plot())
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @jens-siebert, thank you for sharing these detailed observations and code samples with the Ultralytics team! 🚀 This is an automated response to help you get started—an Ultralytics engineer will review and assist you soon.

We recommend reviewing our [Docs](https://docs.ultralytics.com/) for further insights and answers to common questions. For tracking and custom implementation scenarios, the [Tasks](https://docs.ultralytics.com/tasks/) and [Modes](https://docs.ultralytics.com/modes/) sections may be especially useful.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) if possible, as this is invaluable for troubleshooting.

If you have further questions about custom training or tracking behaviors, please include as much information as possible, such as example images, relevant logs, and verification that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community for more support and discussion:
- Real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧
- In-depth discussions: [Discourse](https://community.ultralytics.com/)
- Community threads: [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Please ensure you are using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in these up-to-date, verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for your input and for helping improve Ultralytics YOLO!

**Comment by Y-T-G** [comment]
I guess it's because the update isn't called when there are no detections in Ultralytics, so the "age" of the tracklet doesn't increase.

https://github.com/ultralytics/ultralytics/blob/487e27639595047cff8775dab5e2ff268d8647c4/ultralytics/trackers/track.py#L95-L96

But it increases in your code because you're calling it regardless of detections.

**Comment by ghost** [comment]
Thanks for the quick reply!

I wonder if this is a bug in `model.track` or not... I mean, is the behavior I depicted in the first case above the intended behavior?

**Comment by Y-T-G** [comment]
I guess having the age increase is more correct

**Comment by ghost** [comment]
Okay... Does that mean that the "Bug" label should be attached to this issue so that it's going to be fixed in Ultralytics or shall I fix it on my side of the code?

**Comment by Y-T-G** [comment]
You can open a PR with the fix

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in PR #20945 with great contributions from several community members. This update improves object tracking by ensuring the tracker continues to update even if no detections are found in a frame. By removing the condition that previously skipped tracker updates during such frames, the tracker can now maintain object identities better, especially in scenarios where objects temporarily disappear due to occlusions or missed detections. Users should notice smoother and more reliable tracking results in challenging video sequences.

To test if these changes resolve the issue, it's possible to install the latest code from the main branch using:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
Alternatively, waiting for the next official PyPI release is an option.

Feedback on whether this update resolves the problem would be greatly appreciated. Thank you for reporting this issue, and please share any further observations if the problem persists!
